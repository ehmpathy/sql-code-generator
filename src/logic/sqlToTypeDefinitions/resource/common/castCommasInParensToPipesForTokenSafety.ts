export const castCommasInParensToPipesForTokenSafety = ({ sql }: { sql: string }) => {
  // track state of whether we are in parenthesizes or not
  let areInsideParensDepth = 0; // init as 0, since not inside parens
  const areInsideParens = () => areInsideParensDepth > 0; // whenever areInsideParensDepth > 0, we are inside parens

  // init object to build up the "casted" string
  let castedString = '';

  // loop through each char and update state / replace char as needed
  [...sql].forEach((char) => {
    // 1. manage "areInsideParens" state
    if (char === '(') {
      areInsideParensDepth += 1; // one parens deeper!
    }
    if (char === ')') {
      if (!areInsideParens) throw new Error('exiting parenthesizes without having opened them; not supported'); // fail fast for unsupported case
      areInsideParensDepth -= 1; // mark that we've dropped off one parens now
    }

    // 2. cast the char, considering whether we're in parens or not
    const castedChar = areInsideParens() && char === ',' ? '|' : char; // cast "," if we are inside parens

    // 3. add casted char to string
    castedString += castedChar;
  });

  // return the casted string
  return castedString;
};
