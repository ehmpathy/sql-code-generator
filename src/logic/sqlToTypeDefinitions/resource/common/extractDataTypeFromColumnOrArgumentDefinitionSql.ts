import { DataType } from '../../../../model';

// https://dev.mysql.com/doc/refman/8.0/en/string-types.html
const mysqlStringTypes = [
  'CHAR',
  'VARCHAR',
  'TINYTEXT',
  'TEXT',
  'MEDIUMTEXT',
  'LONGTEXT',
  'ENUM', // note: we treat enums as strings; TODO: https://github.com/uladkasach/sql-code-generator/issues/1
];

// https://www.postgresql.org/docs/9.5/datatype-character.html
const postgresStringTypes = [
  'CHARACTER',
  'CHAR',
  'CHARACTER VARYING',
  'VARCHAR',
  'TEXT',
  'UUID', // this too: https://www.postgresql.org/docs/9.5/datatype-uuid.html
];

// combine the types so we can search over the union
const dbStringTypes = [...mysqlStringTypes, ...postgresStringTypes];

// https://dev.mysql.com/doc/refman/8.0/en/numeric-types.html
const mysqlNumberTypes = [
  'INTEGER',
  'INT',
  'SMALLINT',
  'TINYINT',
  'MEDIUMINT',
  'BIGINT',
  'DECIMAL',
  'NUMERIC',
  'FLOAT',
  'DOUBLE',
];

// https://www.postgresql.org/docs/9.5/datatype-numeric.html
const postgresNumberTypes = [
  'SMALLINT',
  'INT2',
  'INTEGER',
  'INT',
  'INT4',
  'BIGINT',
  'INT8',
  'DECIMAL',
  'NUMERIC',
  'REAL',
  'DOUBLE PRECISION',
  'FLOAT8',
  'SMALLSERIAL',
  'SERIAL2',
  'SERIAL',
  'SERIAL4',
  'BIGSERIAL',
  'SERIAL8',
];

// combine the types so we can search over the union
const dbNumberTypes = [...mysqlNumberTypes, ...postgresNumberTypes];

// https://dev.mysql.com/doc/refman/8.0/en/date-and-time-types.html
const mysqlDateTypes = ['DATE', 'TIME', 'DATETIME', 'TIMESTAMP', 'YEAR'];

// https://www.postgresql.org/docs/9.5/datatype-datetime.html
const postgresDateTypes = [
  'TIMESTAMP',
  'TIMESTAMPTZ',
  'TIMESTAMP WITH TIMEZONE',
  'DATE',
  'TIME',
  'TIMETZ',
  'TIME WITH TIMEZONE',
];

// combine the types so we can search over the union
const dbDateTypes = [...mysqlDateTypes, ...postgresDateTypes];

// https://dev.mysql.com/doc/refman/8.0/en/binary-varbinary.html
const mysqlBinaryTypes = ['BINARY', 'VARBINARY'];

// https://www.postgresql.org/docs/9.5/datatype-binary.html
const postgresBinaryTypes = ['BYTEA'];

// combine the types so we can search over the union
const dbBinaryTypes = [...mysqlBinaryTypes, ...postgresBinaryTypes];

export const extractDataTypeFromColumnOrArgumentDefinitionSql = ({ sql }: { sql: string }) => {
  // 1. cast to upper
  const sqlUpper = sql.toUpperCase();

  // 2. one by one try to see which datatype is being defined in this definition
  if (dbStringTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.STRING;
  if (dbNumberTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.NUMBER;
  if (dbDateTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.DATE;
  if (dbBinaryTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.BUFFER;
  throw new Error(`could not extract data type from '${sql}'`);
};
