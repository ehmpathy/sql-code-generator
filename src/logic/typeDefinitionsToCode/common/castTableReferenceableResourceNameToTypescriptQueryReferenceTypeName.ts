import { ResourceType } from '../../../model';
import { castResourceNameToTypescriptTypeName } from './castResourceNameToTypescriptTypeName';

export const castTableReferenceableResourceNameToTypescriptQueryReferenceTypeName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  // 0. throw error if not view or table, only view and table are possible table references
  if (![ResourceType.TABLE, ResourceType.VIEW].includes(resourceType)) {
    throw new Error('not a table referenceable query');
  }

  // 1. get the regular typescript type name
  const typescriptTypeName = castResourceNameToTypescriptTypeName({ name, resourceType });

  // 2. return the referenceable name
  if (resourceType === ResourceType.TABLE) return typescriptTypeName.replace(/^SqlTable/, 'SqlTableOrView');
  if (resourceType === ResourceType.VIEW) return typescriptTypeName.replace(/^SqlView/, 'SqlTableOrView');
  throw new Error('unexpected'); // this should never occur, but fail fast
};
