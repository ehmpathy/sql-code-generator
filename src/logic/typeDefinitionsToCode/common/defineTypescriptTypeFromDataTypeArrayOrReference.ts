import {
  DataType,
  TypeDefinitionReference,
  TypeDefinition,
  TypeDefinitionOfQueryTableReference,
} from '../../../domain';
import { defineTypescriptTypeFromDataTypeArray } from './defineTypescriptTypeFromDataTypeArray';
import { defineTypescriptTypeFromReference } from './defineTypescriptTypeFromReference/defineTypescriptTypeFromReference';

export const defineTypescriptTypeFromDataTypeArrayOrReference = ({
  type,
  plural,
  queryTableReferences,
  typeDefinitions,
}: {
  type: DataType[] | TypeDefinitionReference;
  plural: boolean;
  typeDefinitions: TypeDefinition[];
  queryTableReferences: TypeDefinitionOfQueryTableReference[];
}) => {
  // if its a type reference, then use that handler
  if (type instanceof TypeDefinitionReference) {
    const defined = defineTypescriptTypeFromReference({
      reference: type,
      queryTableReferences,
      typeDefinitions,
    });
    if (plural) return `${defined}[]`;
    return defined;
  }

  // else, it must be data array. use that handler
  return defineTypescriptTypeFromDataTypeArray({ type });
};
