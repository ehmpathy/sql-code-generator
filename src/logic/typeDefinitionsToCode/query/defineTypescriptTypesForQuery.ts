import { TypeDefinition } from '../../../domain';
import { TypeDefinitionOfQuery } from '../../../domain/objects/TypeDefinitionOfQuery';
import { TypeDefinitionOfQueryInputVariable } from '../../../domain/objects/TypeDefinitionOfQueryInputVariable';
import { castQueryNameToTypescriptTypeName } from '../common/castQueryNameToTypescriptTypeName';
import { defineTypescriptTypeFromDataTypeArrayOrReference } from '../common/defineTypescriptTypeFromDataTypeArrayOrReference';
import { defineTypescriptTypeFromReference } from '../common/defineTypescriptTypeFromReference/defineTypescriptTypeFromReference';

export const defineTypescriptTypesForQuery = ({
  definition,
  allDefinitions,
}: {
  definition: TypeDefinitionOfQuery;
  allDefinitions: TypeDefinition[];
}) => {
  // define the typescript type name for the query
  const typescriptTypeName = castQueryNameToTypescriptTypeName({
    name: definition.name,
  });

  // define the input interface
  const inputVariableNameToTypeDefinitionsMap =
    definition.inputVariables.reduce((summary, thisDefinition) => {
      if (!summary[thisDefinition.name]) summary[thisDefinition.name] = [];
      summary[thisDefinition.name]!.push(thisDefinition);
      return summary;
    }, {} as Record<string, TypeDefinitionOfQueryInputVariable[]>);
  const typescriptInputInterfacePropertyDefinitions = Object.keys(
    inputVariableNameToTypeDefinitionsMap,
  ).map((inputVariableName) => {
    // grab all the defs for this input variable (there can be more than one, because an input variable can be used more than once in the same query. e.g., `WHERE x is null OR x = y`)
    const defs = inputVariableNameToTypeDefinitionsMap[inputVariableName]!;

    // cast each of those to their typescript-type
    const typescriptTypesForReference = defs.map((def) =>
      defineTypescriptTypeFromDataTypeArrayOrReference({
        type: def.type,
        plural: def.plural,
        queryTableReferences: definition.tableReferences,
        typeDefinitions: allDefinitions,
      }),
    );

    // and define the typescript property definition for this input variable as a union of all of those types; // TODO(#50): support intersection; https://github.com/uladkasach/sql-code-generator/issues/50
    return `${inputVariableName}: ${typescriptTypesForReference.join(' | ')};`;
  });
  const typescriptInputInterfaceDefinition =
    typescriptInputInterfacePropertyDefinitions.length // if no inputs, then interface; else, type = null
      ? `
export interface ${typescriptTypeName}Input {
  ${typescriptInputInterfacePropertyDefinitions.join('\n  ')}
}
  `.trim()
      : `
export type ${typescriptTypeName}Input = null;
`.trim();

  // define the output interface
  const typescriptOutputInterfacePropertyDefinitions =
    definition.selectExpressions.map((selectExpression) => {
      const typescriptTypeForReference = defineTypescriptTypeFromReference({
        reference: selectExpression.typeReference,
        queryTableReferences: definition.tableReferences,
        typeDefinitions: allDefinitions,
      });
      return `${selectExpression.alias}: ${typescriptTypeForReference};`;
    });
  const typescriptOutputInterfaceDefinition = `
export interface ${typescriptTypeName}Output {
  ${typescriptOutputInterfacePropertyDefinitions.join('\n  ')}
}
  `.trim();

  // return typescript types
  return `
${typescriptInputInterfaceDefinition}
${typescriptOutputInterfaceDefinition}
  `.trim();
};
