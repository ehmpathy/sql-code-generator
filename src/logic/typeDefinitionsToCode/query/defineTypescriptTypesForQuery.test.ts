import {
  TypeDefinitionOfResourceFunction,
  TypeDefinitionOfResourceTable,
  TypeDefinitionOfResourceView,
} from '../../../domain';
import { extractSqlFromFile } from '../../common/extractSqlFromFile';
import { extractTypeDefinitionFromQuerySql } from '../../sqlToTypeDefinitions/query/extractTypeDefinitionFromQuerySql';
import { defineTypescriptTypesForQuery } from './defineTypescriptTypesForQuery';

describe('defineTypescriptTypesForQuery', () => {
  it('should be able to determine types accurately an example of selecting columns a single table by id', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../__test_assets__/queries/find_image_by_id.sql`,
    });
    const def = extractTypeDefinitionFromQuerySql({
      name: 'find_image_by_id',
      path: '__PATH__',
      sql,
    });
    const code = defineTypescriptTypesForQuery({
      definition: def,
      allDefinitions: [
        new TypeDefinitionOfResourceTable({ name: 'image', columns: [] }),
      ],
    });
    expect(code).toMatchSnapshot();
  });
  it('should be able to determine types accurately when selecting columns from multiple tables, no input variables', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../__test_assets__/queries/select_suggestion.sql`,
    });
    const def = extractTypeDefinitionFromQuerySql({
      name: 'select_suggestion',
      path: '__PATH__',
      sql,
    });
    const code = defineTypescriptTypesForQuery({
      definition: def,
      allDefinitions: [
        new TypeDefinitionOfResourceTable({ name: 'suggestion', columns: [] }),
        new TypeDefinitionOfResourceTable({
          name: 'suggestion_version',
          columns: [],
        }),
      ],
    });
    expect(code).toMatchSnapshot();
  });
  it('should be able to determine types accurately an upsert query', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../__test_assets__/queries/upsert_suggestion.sql`,
    });
    const def = extractTypeDefinitionFromQuerySql({
      name: 'upsert_suggestion',
      path: '__PATH__',
      sql,
    });
    const code = defineTypescriptTypesForQuery({
      definition: def,
      allDefinitions: [],
    });
    expect(code).toMatchSnapshot();
  });
  it('should be able to determine types accurately for this example that selects from tables, functions, and views', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../__test_assets__/queries/find_users_by_last_name.sql`,
    });
    const def = extractTypeDefinitionFromQuerySql({
      name: 'find_users_by_last_name',
      path: '__PATH__',
      sql,
    });
    const code = defineTypescriptTypesForQuery({
      definition: def,
      allDefinitions: [
        new TypeDefinitionOfResourceTable({ name: 'user', columns: [] }),
        new TypeDefinitionOfResourceTable({ name: 'phone', columns: [] }),
        new TypeDefinitionOfResourceView({
          name: 'view_user_profile_current',
          selectExpressions: [],
          tableReferences: [],
        }),
      ],
    });
    expect(code).toMatchSnapshot();
  });
  it('should be able to determine types accurately for this other example that selects from tables, functions, and views', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../__test_assets__/queries/find_users_by_last_name_in.sql`,
    });
    const def = extractTypeDefinitionFromQuerySql({
      name: 'find_users_by_last_name_in',
      path: '__PATH__',
      sql,
    });
    const code = defineTypescriptTypesForQuery({
      definition: def,
      allDefinitions: [
        new TypeDefinitionOfResourceTable({ name: 'user', columns: [] }),
        new TypeDefinitionOfResourceTable({ name: 'phone', columns: [] }),
        new TypeDefinitionOfResourceView({
          name: 'view_user_profile_current',
          selectExpressions: [],
          tableReferences: [],
        }),
      ],
    });
    expect(code).toContain(`candidateLastNames: SqlTableUser['last_name'][]`);
    expect(code).toMatchSnapshot();
  });
  it('should be able to define types accurately a query with subqueries and unnesting', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../__test_assets__/queries/find_train_by_id.sql`,
    });
    const def = extractTypeDefinitionFromQuerySql({
      name: 'find_train_by_id',
      path: '__PATH__',
      sql,
    });
    const code = defineTypescriptTypesForQuery({
      definition: def,
      allDefinitions: [
        new TypeDefinitionOfResourceView({
          name: 'view_train_current',
          selectExpressions: [],
          tableReferences: [],
        }),
        new TypeDefinitionOfResourceTable({ name: 'locomotive', columns: [] }),
        new TypeDefinitionOfResourceTable({ name: 'carriage', columns: [] }),
        new TypeDefinitionOfResourceTable({
          name: 'train_engineer',
          columns: [],
        }),
      ],
    });
    expect(code).toMatchSnapshot();
  });
  it('should be able to define types accurately for a query which selects from the table output of a function', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../__test_assets__/queries/upsert_jerb.sql`,
    });
    const def = extractTypeDefinitionFromQuerySql({
      name: 'upsert_jerb',
      path: '__PATH__',
      sql,
    });
    const code = defineTypescriptTypesForQuery({
      definition: def,
      allDefinitions: [
        new TypeDefinitionOfResourceFunction({
          name: 'upsert_jerb',
          inputs: [],
          output: new TypeDefinitionOfResourceTable({
            name: 'function.output',
            columns: [],
          }),
        }),
      ],
    });
    expect(code).toMatchSnapshot();
  });
  it('should be able to define types correctly for a query which uses an input variable in an or condition (union)', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../__test_assets__/queries/find_all_chat_messages_by_thread.lessthan_or_equalsto_or_null.sql`,
    });
    const def = extractTypeDefinitionFromQuerySql({
      name: 'find_all_chat_messages_by_thread',
      path: '__PATH__',
      sql,
    });
    const code = defineTypescriptTypesForQuery({
      definition: def,
      allDefinitions: [
        new TypeDefinitionOfResourceTable({ name: 'chat_thread', columns: [] }),
        new TypeDefinitionOfResourceView({
          name: 'view_chat_message_current',
          selectExpressions: [],
          tableReferences: [],
        }),
      ],
    });
    expect(code).toContain(
      "until: null | SqlViewViewChatMessageCurrent['created_at'];",
    ); // should have union-ed the two types
    console.log(code);
    expect(code).toMatchSnapshot();
  });
});
