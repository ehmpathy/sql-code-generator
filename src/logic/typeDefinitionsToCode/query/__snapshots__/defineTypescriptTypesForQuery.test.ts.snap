// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineTypescriptTypesForQuery should be able to define types accurately a query with subqueries and unnesting 1`] = `
"export interface SqlQueryFindTrainByIdInput {
  id: SqlViewViewTrainCurrent['id'];
}
export interface SqlQueryFindTrainByIdOutput {
  id: SqlViewViewTrainCurrent['id'];
  uuid: SqlViewViewTrainCurrent['uuid'];
  home_station_geocode: SqlFunctionJsonBuildObjectOutput;
  combination_id: SqlViewViewTrainCurrent['combination_id'];
  locomotive_uuids: SqlFunctionArrayAggOutput;
  carriage_uuids: SqlFunctionArrayAggOutput;
  engineer_uuids: SqlFunctionArrayAggOutput;
  status: SqlViewViewTrainCurrent['status'];
}"
`;

exports[`defineTypescriptTypesForQuery should be able to define types accurately for a query which selects from the table output of a function 1`] = `
"export interface SqlQueryUpsertJerbInput {
  ownerUuid: SqlFunctionUpsertJerbInput['0'];
  title: SqlFunctionUpsertJerbInput['1'];
  details: SqlFunctionUpsertJerbInput['2'];
  dueDate: SqlFunctionUpsertJerbInput['3'];
}
export interface SqlQueryUpsertJerbOutput {
  id: SqlFunctionUpsertJerbOutput['id'];
  uuid: SqlFunctionUpsertJerbOutput['uuid'];
}"
`;

exports[`defineTypescriptTypesForQuery should be able to define types correctly for a query which uses an input variable in an or condition (union) 1`] = `
"export interface SqlQueryFindAllChatMessagesByThreadInput {
  threadUuid: SqlTableChatThread['uuid'];
  until: null | SqlViewViewChatMessageCurrent['created_at'];
  limit: number;
}
export interface SqlQueryFindAllChatMessagesByThreadOutput {
  id: SqlViewViewChatMessageCurrent['id'];
  uuid: SqlViewViewChatMessageCurrent['uuid'];
  thread_uuid: SqlTableChatThread['uuid'];
  sent_at: SqlViewViewChatMessageCurrent['sent_at'];
  text: SqlViewViewChatMessageCurrent['text'];
}"
`;

exports[`defineTypescriptTypesForQuery should be able to determine types accurately an example of selecting columns a single table by id 1`] = `
"export interface SqlQueryFindImageByIdInput {
  id: SqlTableImage['id'];
}
export interface SqlQueryFindImageByIdOutput {
  id: SqlTableImage['id'];
  url: SqlTableImage['url'];
  caption: SqlTableImage['caption'];
  alt_text: SqlTableImage['alt_text'];
  credit: SqlTableImage['credit'];
}"
`;

exports[`defineTypescriptTypesForQuery should be able to determine types accurately an upsert query 1`] = `
"export interface SqlQueryUpsertSuggestionInput {
  suggestionSource: SqlFunctionUpsertSuggestionInput['0'];
  externalId: SqlFunctionUpsertSuggestionInput['1'];
  suggestedIdeaId: SqlFunctionUpsertSuggestionInput['2'];
  status: SqlFunctionUpsertSuggestionInput['3'];
  result: SqlFunctionUpsertSuggestionInput['4'];
}
export interface SqlQueryUpsertSuggestionOutput {
  id: SqlFunctionUpsertSuggestionOutput;
}"
`;

exports[`defineTypescriptTypesForQuery should be able to determine types accurately for this example that selects from tables, functions, and views 1`] = `
"export interface SqlQueryFindUsersByLastNameInput {
  lastName: SqlTableUser['last_name'];
}
export interface SqlQueryFindUsersByLastNameOutput {
  id: SqlTableUser['id'];
  full_name: SqlFunctionConcatOutput;
  age: SqlTableUser['age'];
  phone_number: SqlTablePhone['number'];
  description: SqlViewViewUserProfileCurrent['description'];
}"
`;

exports[`defineTypescriptTypesForQuery should be able to determine types accurately for this other example that selects from tables, functions, and views 1`] = `
"export interface SqlQueryFindUsersByLastNameInInput {
  candidateLastNames: SqlTableUser['last_name'][];
}
export interface SqlQueryFindUsersByLastNameInOutput {
  id: SqlTableUser['id'];
  full_name: SqlFunctionConcatOutput;
  age: SqlTableUser['age'];
  phone_number: SqlTablePhone['number'];
  description: SqlViewViewUserProfileCurrent['description'];
}"
`;

exports[`defineTypescriptTypesForQuery should be able to determine types accurately when selecting columns from multiple tables, no input variables 1`] = `
"export type SqlQuerySelectSuggestionInput = null;
export interface SqlQuerySelectSuggestionOutput {
  id: SqlTableSuggestion['id'];
  uuid: SqlTableSuggestion['uuid'];
  suggestion_source: SqlTableSuggestion['suggestion_source'];
  external_id: SqlTableSuggestion['external_id'];
  suggested_idea_id: SqlTableSuggestion['suggested_idea_id'];
  resultant_curated_idea_id: SqlTableSuggestion['resultant_curated_idea_id'];
  status: SqlTableSuggestionVersion['status'];
  result: SqlTableSuggestionVersion['result'];
  created_at: SqlTableSuggestion['created_at'];
  effective_at: SqlTableSuggestionVersion['effective_at'];
  updated_at: SqlTableSuggestionVersion['created_at'];
}"
`;
